File: dsa-pros/leetcode/array/JavaScript/timiBello/Week 6/wk6.ts
//Single Number
function singleNumber(nums: number[]): number {
let result = 0;
  for (let num of nums) {
    result ^= num;
  }
  return result;
};


//Majority Element II/ using hash map
function majorityElement(nums: number[]): number[] {
    const result = [];
  const frequencyMap = {};

  // Step 1: Count frequencies
  for (let i = 0; i < nums.length; i++) {
    if (frequencyMap[nums[i]] === undefined) {
      frequencyMap[nums[i]] = 1;
    } else {
      frequencyMap[nums[i]]++;
    }
  }

  // Step 2: Check frequencies
  const threshold = Math.floor(nums.length / 3);
  for (const num in frequencyMap) {
    if (frequencyMap[num] > threshold) {
      result.push(parseInt(num));
    }
  }

  return result;
};


/*
JS(using a map(numsindices)
**/
var twoSum = function(nums, target) {
    const numIndices = new Map();

  for (let i = 0; i < nums.length; i++) {
    const complement = target - nums[i];

    if (numIndices.has(complement)) {
      return [numIndices.get(complement), i];
    }

    numIndices.set(nums[i], i);
  }

  // No solution found
  return [];

};

//Container With Most Water(two pointer approach) 
//JS(two pointer approach)
var maxArea = function(height) {
    let left = 0;
  let right = height.length - 1;
  let maxArea = 0;

  while (left < right) {
    const currentArea = Math.min(height[left], height[right]) * (right - left);
    maxArea = Math.max(maxArea, currentArea);

    if (height[left] < height[right]) {
      left++;
    } else {
      right--;
    }
  }

  return maxArea;
};

File: dsa-pros/leetcode/array/JavaScript/timiBello/Week 6/wk6.py
#Single Number(use the XOR (exclusive OR) operation. 
# The XOR of two equal numbers is 0, and the XOR of a number with 0 is the number itself. By XORing all the elements in the array, the duplicate elements will cancel out, and the remaining element will be the single one)

class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        result = 0
        for num in nums:
            result ^= num
        return result


#Majority Element II
class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        count = {}
        result = []

    # Count occurrences of each element
        for num in nums:
            count[num] = count.get(num, 0) + 1

    # Check if element occurs more than n/3 times
        threshold = len(nums) // 3
        for num, freq in count.items():
            if freq > threshold:
                result.append(num)

        return result


#Two Sum(two pointer approach) 
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        num_indices = {}
        
        
        for i, num in enumerate(nums):
            complement = target - num
            
            if complement in num_indices:
                return [num_indices[complement], i]
            
            num_indices[num] = i
    
    # No solution found
        return []


#Container With Most Water(two pointer approach) 
class Solution(object):
    def maxArea(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        left = 0
        right = len(height) - 1
        max_area = 0
        
        while left < right:
            current_area = min(height[left], height[right]) * (right - left)
            max_area = max(max_area, current_area)
            
            
            if height[left] < height[right]:
                left += 1
            else:
                right -= 1
        
        return max_area